aten/src/ATen/core/LazyTensor.cpp

```diff
@@ -0,0 +1,317 @@
+      } else {
+        g->setInsertPoint(dep_node);
+        torch::jit::Value* v;
+        if (ival.isNone()) {
+          v = g->insertNode(g->createNone())->output();
+        } else {
+          v = insertConstant(*g, ival);
+        }
+        dep_node->addInput(v);
+      }
+    }
+
+    frontier = next_frontier;
+  }
+
+  return std::make_pair(output, inputs);
+}
+
+struct Item {
+  Item(std::pair<std::string, size_t> var) : tag(VAR), var_(var) {}
+  Item(c10::IValue val) : tag(CONST), val_(val) {}
+
+  enum { VAR, CONST } tag;
+
+  const std::pair<std::string, size_t>& var() const {
+    TORCH_INTERNAL_ASSERT(tag == VAR);
+    return var_;
+  }
+  const c10::IValue& val() const {
+    TORCH_INTERNAL_ASSERT(tag == CONST);
+    return val_;
+  }
+
+  // union {
+  std::pair<std::string, size_t> var_ = std::make_pair("", 0);
+  c10::IValue val_;
+  //};
+  Item(Item&&) = default;
+  ~Item() {}
+};
+
+using Key = std::vector<Item>;
+
+bool keyEqual(const Key& k1, const Key& k2) {
+  if (k1.size() != k2.size()) {
+    return false;
+  }
+  for (auto i = 0; i < k1.size(); ++i) {
+    const auto& a = k1[i];
+    const auto& b = k2[i];
+    if (a.tag != b.tag) {
+      return false;
+    }
+    if (a.tag == Item::VAR) {
+      if (a.var() != b.var()) {
+        return false;
+      }
+    } else {
+      // TODO other IValue checks
+      if (a.val().isInt() && b.val().isInt() &&
+          a.val().toInt() == b.val().toInt()) {
+        return true;
+      }
+      if (a.val().isBool() && b.val().isBool() &&
+          a.val().toBool() == b.val().toBool()) {
+        return true;
+      }
+      if (a.val().isDouble() && b.val().isDouble() &&
+          a.val().toDouble() == b.val().toDouble()) {
+        return true;
+      }
+      if (a.val().isTensor() && b.val().isTensor()) {
+        // TODO type check
+        return true;
+      }
+      return false;
+    }
+  }
+  return true;
+}
+
+struct GraphCache {
+  std::shared_ptr<torch::jit::GraphExecutor> get(const Key& k) {
+    for (const auto& pair : gs_) {
+      if (keyEqual(pair.first, k)) {
+        return pair.second;
+      }
+    }
+    return std::shared_ptr<torch::jit::GraphExecutor>(nullptr);
+  }
+
+  void put(Key&& k, std::shared_ptr<torch::jit::GraphExecutor> g) {
+    gs_.emplace_back(std::make_pair(std::move(k), g));
+  }
+
+  std::list<std::pair<Key, std::shared_ptr<torch::jit::GraphExecutor>>> gs_;
+};
+
+static GraphCache graph_cache;
+
+std::pair<
+    std::shared_ptr<torch::jit::GraphExecutor>,
+    std::vector<const LazyTensorImpl*>>
+getGraph(const LazyTensorImpl* lt_) {
+  // Construct a straight line repr of the LT
+  //  a = ?.to_lazy()
+  //  b = ?.to_lazy()
+  //  c = a @ b
+  //  d = c * 2
+  // becomes
+  //  *[0], @[0], 2, a, b
+  // that is, all LTs with exprs are "ExprName[index]", constants are constant
+  // and LTs with tensor bindings are LT*
+  // We can theoretically do memory comparison on these
+
+  std::vector<Item> record;
```

Comment:
This chunk of code can be made into a function like `hashgraph` or something. It produces a vector of records.